src/libtoast/src/toast_tod_filter.cpp

```diff
@@ -52,108 +58,228 @@ void toast::filter_polynomial(int64_t order, size_t n, uint8_t * flags,
 
             double dx = 2. / scanlen;
             double xstart = 0.5 * dx - 1;
-
-            // FIXME: can we be more clever here and remove the conditionals
-            // inside the for loop?
-            #pragma omp simd
-            for (int64_t i = 0; i < scanlen; ++i) {
-                double x = xstart + i * dx;
-                int64_t offset = i * norder;
-                if (norder > 0) full_templates[offset++] = 1;
-                if (norder > 1) full_templates[offset++] = x;
-                for (int64_t iorder = 1; iorder < norder - 1; ++iorder) {
-                    full_templates[offset] =
-                        ((2 * iorder + 1) * x * full_templates[offset - 1]
-                         - iorder * full_templates[offset - 2]) / (iorder + 1);
-                    ++offset;
+            double * current, * last, * lastlast;
+
+            for (size_t iorder = 0; iorder < norder; ++iorder) {
+                current = &full_templates[iorder * scanlen];
+                if (iorder == 0) {
+                    #pragma omp simd
+                    for (size_t i = 0; i < scanlen; ++i) current[i] = 1;
+                } else if (iorder == 1) {
+                    #pragma omp simd
+                    for (size_t i = 0; i < scanlen; ++i) {
+                        const double x = xstart + i * dx;
+                        current[i] = x;
+                    }
+                } else {
+                    last = &full_templates[(iorder - 1) * scanlen];
+                    lastlast = &full_templates[(iorder - 2) * scanlen];
+                    double orderinv = 1. / iorder;
+                    #pragma omp simd
```

Comment:
Note that `omp simd` assumes all allocated memory buffers are aligned, and will return incorrect results if not.  In this case the `last` and `lastlast` are pointers to aligned memory, so it is ok.

Reply:
Great!